### DISCUSSION TITLE
What is the globbing pattern to exclude folders starting with a '.' or '_'?

### DISCUSSION THREAD

**Discussion by AceTheWiz** [statiqdev/Discussions#11] [Q&A]
What is the pattern to exclude folders starting with a '.' or '_'? 

I tried this but I can't tell if its working. (because of multiple other issues I'm having).
```C#
.AddSetting(WebKeys.AssetFiles, "{!.,!_,}**/*")
```

**Comment by daveaglick** [comment]
That looks right at first glance. You might need to put the exclusion pattern after a recursive wildcard, something like:

`**/{!.,!_,}*/**/*`

I’ve been meaning to add a globbing tester for a while, might try to get that into the next release. In other words, given a root folder and a pattern, list the files that would match. Maybe a second mode where you provide an example path and it says whether that would match. That should make debugging globbing patterns a lot easier.

**Reply by daveaglick** [reply]
Let me know if I can help with the other problems you’re having

**Reply by daveaglick** [reply]
FYI, I just added `glob test` and `glob eval` commands to help debug globbing patterns. I'll document these new commands on the site after they ship in the next release of Statiq Framework.

**Reply by AceTheWiz** [reply]
> That looks right at first glance. You might need to put the exclusion pattern after a recursive wildcard, something like:
> 
> `**/{!.,!_,}*/**/*`
> 
> I’ve been meaning to add a globbing tester for a while, might try to get that into the next release. In other words, given a root folder and a pattern, list the files that would match. Maybe a second mode where you provide an example path and it says whether that would match. That should make debugging globbing patterns a lot easier.

If only I read your reply properly I would have got it right much quicker. 🤦

This is what I found works. ```**/{!.*,!~*,}/**/*``` otherwise the ```{!.,!_,}*``` excludes the files in the top level folder.

**Comment by JohnGoldsmith** [comment]
Hi,  

I thought I'd add to this thread as the [`glob eval`](https://www.statiq.dev/framework/execution/files#globbing) command helped me a lot.

I found that I wasn't clear on how the patterns gets split up and so I used Linqpad to dump the [`expandedPatterns`](https://github.com/statiqdev/Statiq.Framework/blob/0c192e4b2b9f7434b8f244ce36bfeb3f670d8da4/src/core/Statiq.Common/IO/Globbing/Globber.cs#L55) which was helpful in understanding what `Globber` is doing and also what was wrong with my own pattern (trying to ignore underscore prefixed folders).

So a straight forward pattern returns all of the paths:

![image](https://user-images.githubusercontent.com/2076297/136594130-46ebd1bd-411d-4dd5-98ba-01e61f55c758.png)

...and one with a more complex pattern returns the expected output but dumping the `expandedPatterns` really helps:

![image](https://user-images.githubusercontent.com/2076297/136594093-59ab5f45-b11a-4cbf-b18a-f2d8e305c72b.png)

I wonder if it would be useful to include the `expandedPatterns` output as part of the `glob eval` command?

Best regards

John

**Reply by daveaglick** [reply]
I love this idea - added https://github.com/statiqdev/Statiq.Framework/issues/206 to track it
